examples/AcousticWave2D_codegen.py

```diff
@@ -95,21 +95,43 @@ def _init_taylor(self, nt):
         # domain to the border
 
         # Forward wave equation
+
+        def damp_boundary(damp):
+            h = self.h
+            dampcoeff = 1.5 * np.log(1.0 / 0.001) / (40 * h)
+            nx, ny = self.model.get_shape()
+            nbpml = self.nbpml
+            for i in range(nbpml):
+                pos = np.abs((nbpml-i)/nbpml)
+                val = dampcoeff * (pos - np.sin(2*np.pi*pos)/(2*np.pi))
+                damp[i, :] += val
+                damp[-i, :] += val
+                damp[:, i] += val
+                damp[:, -i] += val
+
         wave_equation = m * dtt - (dxx + dyy) + e * dt
         stencil = solve(wave_equation, p(x, y, t+s))[0]
         self.nt = nt
-
+        m_sub = DenseData("m", self.model.vp.shape, self.dtype)
+        m_sub.initializer = lambda ref: np.copyto(ref, self.model.vp**(-2))
```

Comment:
This really sounds like premature optimisation to me! Delaying the initialisation does not change the fact that a certain numpy operation copies data (or not), it just adds confusion as to when that happens. But essentially we are are still executing the same operation (with or without the `np.copyto`). The point we discussed was the notion of particular values being computed on-the-fly, ie. in the code-generated loop and possibly from stencil objects, but this is not the case here. And I would argue that we can add delayed initialisation once we actually have an example case where we need this, but without an explicit test this is simply over-engineering.

Also, looking at this again, I don't think `m_sub.data = self.model.vp**-2` will do a pointer assignment, since `self.model.vp**-2` is never stored in an explicit array anywhere, or am I missing something here?